iceoryx_hoofs/source/posix_wrapper/named_semaphore.cpp

```diff
@@ -0,0 +1,189 @@
+                return cxx::error<SemaphoreError>(SemaphoreError::OUT_OF_MEMORY);
+            default:
+                LogError() << "This should never happen. An unknown error occurred while opening the semaphore \""
+                           << m_name << "\".";
+                return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+            }
+        }
+        else
+        {
+            if (m_openMode == OpenMode::OPEN_EXISTING && result->errnum == ENOENT)
+            {
+                LogError() << "Unable to open semaphore since no semaphore with the name \"" << m_name << "\" exists.";
+                return cxx::error<SemaphoreError>(SemaphoreError::NO_SEMAPHORE_WITH_THAT_NAME_EXISTS);
+            }
+            else if (result->errnum != EINVAL && result->errnum != ENOENT)
+            {
+                uninitializedSemaphore.emplace(result.value().value, m_name, hasOwnership);
+                return cxx::success<>();
+            }
+        }
+    }
+    /// END: try to open existing semaphore
+
+    /// BEGIN: create semaphore
+    hasOwnership = true;
+    auto result = posixCall(iox_sem_open_ext)(m_name.c_str(),
+                                              convertToOflags(m_openMode),
+                                              static_cast<mode_t>(m_permissions),
+                                              static_cast<unsigned int>(m_initialValue))
+                      .failureReturnValue(IOX_SEM_FAILED)
+                      .evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EACCES:
+            LogError() << "Insufficient permissions to create semaphore \"" << m_name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::PERMISSION_DENIED);
+        case EEXIST:
+            LogError() << "A semaphore with the name \"" << m_name
+                       << "\" does already exist. This should not happen until there is a race condition when multiple "
+                          "instances try to create the same named semaphore concurrently.";
+            return cxx::error<SemaphoreError>(SemaphoreError::ALREADY_EXIST);
+        case EMFILE:
+            LogError() << "The per-process limit of file descriptor exceeded while creating the semaphore \"" << m_name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENFILE:
+            LogError() << "The system wide limit of file descriptor exceeded while creating the semaphore \"" << m_name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENOMEM:
+            LogError() << "Insufficient memory to create the semaphore \"" << m_name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::OUT_OF_MEMORY);
+        default:
```

Comment:
According to https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_open.html, `ENOSPC` is a possible error too.

Reply:
You are right but for now I would ignore it. We need to define to which posix standard we stick and then all the error messages may have to refined a bit.

I would address this when we are at that point.